Gemcalibration package second PR

cgalloni opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue(s) here. -->
Implementation of the XDAQ web interface

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Code tested on gem904daq02

### Screenshots (if appropriate):

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--- Template thanks to https://www.talater.com/open-source-templates/#/page/99 -->

jsturdy committed:
Update XHAL path

Change expected `xhal` path from `XHALROOT` to `XHAL_ROOT`,
to be consistent with other packages and `etc/profile.d/gemdaqenv.sh`

cgalloni committed:
files checked out from <fac3430>

cgalloni committed:
files checked out from <f312213>

cgalloni committed:
adding dependences for gemcalibration and removed bug in Makefile

mexanick commented on setup/etc/profile.d/gemdaqenv.sh:
```diff
@@ -9,7 +9,7 @@ if [ "$XDAQ_ROOT" != "0" ]; then
     export XDAQ_ROOT=/opt/xdaq
     export XDAQ_DOCUMENT_ROOT=$XDAQ_ROOT/htdocs
     # only do this for regular users
-    # export XDAQ_DOCUMENT_ROOT=/data/xdaq/${USER}
```
This change must be reverted, as it is only required for development and not suitable for releasing

cgalloni replied:
done

jsturdy replied:
This file was *not* listed in my instructions for cleaning up the old PR...

cgalloni committed:
reverting the change in gemdaqenv.sh

jsturdy commented on Makefile:
```diff
@@ -11,6 +11,7 @@ SUBPACKAGES := \
         gemonlinedb \
+        gemcalibration \
         # gemHwMonitor \
 
```
Missing `gemcalibration` rule in top-level `Makefile`

jsturdy commented on config/mfPythonDefsGEM.mk:
```diff
@@ -9,6 +9,10 @@ PYTHON_LIB     ?= python${PYTHON_VERSION}
+PYTHONGLIBS  = $(shell pkg-config python --glibs)
+
+IncludeDirs+=${PYTHON_INCLUDE_PREFIX}
+
```
Where did these get added from?

cgalloni replied:
I was pointed to this https://github.com/cms-gem-daq-project/cmsgemos/blob/generic-amc-release-v3/config/mfPythonDefsGEM.mk because of problems during the compilation related to the file Python.h not found.

mexanick replied:
Actually this has to be changed to python compilation flags because that has been changed towards the develop...
https://github.com/cms-gem-daq-project/cmsgemos/blob/develop/config/mfPythonDefsGEM.mk

jsturdy replied:
This *should* be taken care of by the changes to the `gemcalibration/Makefile`, which was why I was curious why this change was introduced.

```
UserCFlags  +=$(PYTHONCFLAGS)
UserCCFlags +=$(PYTHONCFLAGS)
```

jsturdy commented on gemcalibration/include/gem/calib/CalibrationWeb.h:
```diff
@@ -0,0 +1,91 @@
+
+    class CalibrationWeb: public gem::base::GEMWebApplication
+      {
+        // friend class GEMMonitor;
```
Remove unnecessary commented lines

jsturdy commented on gemcalibration/src/common/Calibration.cc:
```diff
@@ -0,0 +1,226 @@
+    //DACSCAN type
+    if (m_calType==DACSCANV3){
+      std::map<dacScanType_t, std::map<std::string, uint32_t> >::iterator it;
+      for(it=m_dacScanTypeParams.begin();it!=m_dacScanTypeParams.end();it++){
```
* whitespace should follow the form `for (blah) {` (@lmoureaux has added many of my style requirements to the `CONTRIBUTING.md` guidelines)
* Use scoped range based loops with `auto &` or `auto const&` as necessary rather than declaring an iterator outside the loop (unless you have a valid reason not to)

jsturdy commented on gemcalibration/src/common/Calibration.cc:
```diff
@@ -0,0 +1,226 @@
+	checked = cgi.queryCheckbox(t_stream.str());
+	if (checked) {
+	  std::map<std::string, uint32_t> t_dacScan_parameters = it->second;
+	  for (auto dacScan_parameter: t_dacScan_parameters) {
```
again, scope variables unless you have a valid reason not to
```c++
for (auto & par : it->second)
```

jsturdy commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,394 @@
+  *out << "<script src=\"/gemdaq/gembase/html/scripts/bootstrap.js\"></script>" << std::endl;
+  *out << "<script src=\"https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js\"></script>" << std::endl;
+
+  *out << "<div class=\"xdaq-tab-wrapper\">" << std::endl;
```
Formatting in this file *must* be for `c++`, not `html`.
If you also want formatted `html` in the output, ensure what you write to the stream is formatted as desired.

jsturdy commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,394 @@
+                   *out << "<div class=\"col-md-3\">" << std::endl;
+
+                    *out << "<select class=\"form-control form-control-lg\" id=\"cal_type_select\" name=\"cal_type_select\" onChange=\"selectCalType()\">" << std::endl;
+                        *out << "<option disabled selected value> -- select an option -- </option>" << std::endl;
```
Same comment throughout this block re formatting

jsturdy commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,394 @@
+                    << "</div>" << std::endl; // end <div class="row">
+                } // end loop over NSHELF
+            *out << "</div>" << std::endl; // end container
+        *out << "</form>"<< std::endl;
```
This entire file should be reformatted, and subsequently, all trailing whitespace removed, before I can perform further review.

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+//#include "gem/daqmon/DaqMonitor.h"
+
+#define NAMC 12
+#define NSHELF 2
```
For sure, this should not be hardcoded, take it as a parameter from the xml config

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+#include "gem/utils/exception/Exception.h"
+//#include "gem/daqmon/DaqMonitor.h"
+
+#define NAMC 12
```
Use `gem::base::GEMApplication::MAX_AMCS_PER_CRATE`

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+    typedef enum calType calType_t;
+
+
+    enum dacScanType {CFG_CAL_DAC, CFG_BIAS_PRE_I_BIT, CFG_BIAS_PRE_I_BLCC, CFG_BIAS_PRE_I_BSF, CFG_BIAS_SH_I_BFCAS, CFG_BIAS_SH_I_BDIFF, CFG_BIAS_SD_I_BDIFF, CFG_BIAS_SD_I_BFCAS, CFG_BIAS_SD_I_BSF, CFG_BIAS_CFD_DAC_1, CFG_BIAS_CFD_DAC_2, CFG_HYST, CFG_THR_ARM_DAC, CFG_THR_ZCC_DAC, CFG_BIAS_PRE_VREF, CFG_VREF_ADC}; 
```
Likewise, and document

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+        void setCalType(xgi::Input *in, xgi::Output *out)
+          throw (xgi::exception::Exception);
+
+        std::vector<Calibration*> v_gemcal;
```
What is this? I.e., why does the `Calibration` class contain a vector of pointers to `Calibration` items?
Use smart pointers unless impossible due to compatibility with external libraries

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+
+        calType_t m_calType;
+
+        std::map<calType_t, std::map<std::string, uint32_t>> m_scanParams{
```
This needs documentation

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+	  //std::string name;
+	  
+	  std::string label;
+	  std::vector<std::string>  radio_options;
```
is `radio_` necessary?
the struct type already labels it such

cgalloni replied:
changed to simply `options`.

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+	};
+	
+    
+	std::map<std::string, scanParamsRadioSelector> m_scanParamsRadioSelector{
```
Without seeing how these (the various `map` objects herein defined) are used, I would encode these in `enums`.
Documentation *might* help decide whether that is reasonable.
Additionally, that it is labeled as a radio selector, it strongly indicates that it is rather part of the UI (`CalibrationWeb`) rather than part of the core functionality, (`Calibration`)
If they are not converted to appropriate enums, they should almost definitely be `const` or `constexpr`

cgalloni replied:
Radio selectors as part of the CalibrationWeb UI are implemented for those parameters that don't need forms, but just can assume certain values. For instance for the TriggerType, ComparatorType to be used, etc.. These parameters will need to be fed to the back-end code eventually.

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+         */
+        bool isGEMApplication(const std::string& classname) const;
+        xdata::Integer m_shelfID;
+        log4cplus::Logger m_logger; //FIXME should be removed!
```
Then why isn't it?

mexanick replied:
looks like remnant from monitoring package... Indeed has to be removed

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+         * @throws
+         */
+        bool isGEMApplication(const std::string& classname) const;
+        xdata::Integer m_shelfID;
```
what is the purpose of this?

cgalloni replied:
removed

jsturdy commented on gemcalibration/include/gem/calib/CalibrationWeb.h:
```diff
@@ -0,0 +1,91 @@
+/** @file CalibrationWeb.h */
+
+#ifndef GEM_GEMCAL_CALIBRATIONWEB_H
```
header guards should match the namespace, i.e., `GEM_CALIB_` (also elsewhere)

cgalloni replied:
done

jsturdy commented on gemcalibration/src/common/Calibration.cc:
```diff
@@ -0,0 +1,226 @@
+
+void gem::calib::Calibration::init()
+{
+  /*  v_daqmon.clear();
```
Another set of needed to be removed commented lines

jsturdy commented on gemcalibration/src/common/Calibration.cc:
```diff
@@ -0,0 +1,226 @@
+  out->getHTTPResponseHeader().addHeader("Content-Type", "application/json");
+  //this->stopMonitoring();
+  *out << " { \"mon_state\":\"PAUSED\"}" << std::endl;
+  std::cout << "ciao! " << std::endl;
```
Use logger, not `cout`

jsturdy commented on gemcalibration/src/common/Calibration.cc:
```diff
@@ -0,0 +1,226 @@
+    
+    std::stringstream t_stream;
+    t_stream.clear();
+    t_stream.str(std::string());
```
Are these two lines necessary before the stream has ever been used?

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+	  {"latency"   , "Latency (BX)"},
+	  {"trimValues", "Points in dac range (odd)"}, // TODO:need to be implemented properly in the back end in order to get a given number of points {-63,0,63}
+	    };
+        std::map<std::string, uint32_t> amc_optical_links;
```
member variables must use `m_<name>` where `<name>` should utilize "almost" `CamelCase`/`PascalCase` (see updated `CONTRIBUITING.md` guidelines)

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,240 @@
+namespace gem {
+  namespace calib {
+
+    enum calType {NDEF, GBTPHASE, LATENCY, SCURVE, SBITARMDACSCAN, ARMDACSCAN, TRIMDAC, DACSCANV3, CALIBRATEARMDAC}; 
```
I would put these `enums` in a dedicated file (e.g., `GEMCalibEnums.h`).
See, e.g., `gembase/include/gem/base/utils/GEMEnums.h`

bdorney commented:
Is it possible to get a pdf like last time that shows the final interface that this PR creates for each scan type?

cgalloni commented:
Hi Brian, yes,  I'll do that.

cgalloni commented:
Hi @bdorney, meanwhile I implement the other comments see here the current interface. 
[CalibrationInterface2.pdf](https://github.com/cms-gem-daq-project/cmsgemos/files/3108460/CalibrationInterface2.pdf)

cgalloni committed:
moving enums to separate file, taking NSHELF from XML and tentative documentation

cgalloni committed:
removing commented out parts in CalibrationWeb.cc

cgalloni committed:
fix python lib in config/mfPythonDefsGEM.mk

cgalloni committed:
c indentation for code while html indentation for xgi output

jsturdy commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,307 @@
+  *out << "<link  rel=\"stylesheet\" href=\"/gemdaq/gembase/html/css/bootstrap.css\" type=\"text/css\">" << std::endl;
+  *out << "<link  rel=\"stylesheet\" href=\"/gemdaq/gembase/html/css/bootstrap.min.css\" type=\"text/css\">" << std::endl;
+  *out << "<script src=\"/gemdaq/gembase/html/scripts/bootstrap.js\"></script>" << std::endl;
+  *out << "<script src=\"https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js\"></script>" << std::endl;
```
External links don't work on the technical network, please select a version and do as you have done with `bootstrap`

jsturdy commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,307 @@
+
+  for (unsigned int i=0; i< radio_param.options.size();i++) {
+    if (i==0) {
+      *out<<"<div class=\"form-check-input radio-inline\"> <label><input type=\"radio\" name=\""<<paramName<<"\" id=\""<<paramName<<"_radio_"<<i<<"\" value="<<i<<" checked>"<<radio_param.options[i]<<"</label></div>"<< std::endl;
```
If you want to check your HTML indentation, you can load up your page in a browser and then view the source.
Here you've moved the indentation to the beginning of the HTML page, which is not likely what you wanted

jsturdy commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,307 @@
+  *out << "<div class=\"col-md-6\">"<< std::endl;
+  *out << "    <div class=\"form-check\">" << std::endl;
+
+  for (unsigned int i=0; i< radio_param.options.size();i++) {
```
Prefer preincrement unless specifically need a postincremenet

jsturdy commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,307 @@
+  *out << "                <button id=\"dacScan_dropdown_button\" class=\"btn btn-lg btn-outline dropdown-toggle\" data-toggle=\"dropdown\">Settings</button>" << std::endl;
+  *out << "                <div class=\"dropdown-menu pre-scrollable\">" << std::endl;
+
+  for(auto it=dynamic_cast<gem::calib::Calibration*>(p_gemApp)->m_dacScanTypeParams.begin();it!=dynamic_cast<gem::calib::Calibration*>(p_gemApp)->m_dacScanTypeParams.end();it++){
```
You can probably clean this up with 
```c++
for (auto & it : dynamic_cast<gem::calib::Calibration*>(p_gemApp)->m_dacScanTypeParams) {
```

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,244 @@
+    typedef enum calType calType_t;
+    
+    
+    typedef enum dacScanType dacScanType_t;
```
why not also put the typedefs in the enum defs header?

bdorney commented:
> Hi @bdorney, meanwhile I implement the other comments see here the current interface.
> [CalibrationInterface2.pdf](https://github.com/cms-gem-daq-project/cmsgemos/files/3108460/CalibrationInterface2.pdf)

Hi @cgalloni it seems like you addressed a lot of the comments we had the last time on the interface layout.  Thanks a lot.  Another round of comments here.  Apologies if there is some duplication.

What is `CalPulse phase` here supposed to be?  Is this the time between L1A and calpulse? Or is this the phase in the VFAT3 calibration unit? e.g. `GEM_AMC.OH.OHX.GEB.VFATY.CFG_CAL_PHI`?

Now slide specific comments.

# Slide 3

- A GBT Phase scan does not need a trigger source, drop this.
- Add the following additional fields:
   - phase min (int) defaulting to 0,
   - phase max (int) defaulting to **14** (we've seen phase 15 is always bad, so no reason to include it)
   - step size (int) defaulting to 1

Also the yellow bar says `HELLOW WORLD`.  I guess this is not intended?

# Slide 4

- Change default `ohMask` to `0xfff` since we should assume in P5 we will have 12 OH's per CTP7; or default it to `0xffc` since on QC8 `OH0` and `OH1` are not used per CTP7.  But the `0x3ff` is a legacy from the slice test.

# Slide 5

- Change default `Throttle` to `0`; our use case in near term is entirely cosmic data taking or internal cal pulse and for cosmic I fear having a default of 100 will cause the scans to take 100 times as longer :D
- Change default `CFG_THR_ARM_DAC` to at least 50 or 100.  Having a default of 0 will probably risk to end up recording only noise.

If the `calibration pulse` radial button is selected then the `Pulse Delay` field (e.g. from slide 6) should be present here.  It's meaningless for the `particle` option.

Here I assume `Pulse Delay` is the difference between the calpulse and the L1A and this should be configurable for a latency scan with the calibration pulse."

# Slide 6

- What's the rational in including a field for `CFG_THR_ARM_DAC` on slide 5 for the latency scan but not here?  Just wondering if we should be adding/dropping this; Is the use case the user configures with some known config in the DB? Or makes a config by hand and then makes a test?  Or...?  If a known config from the DB is to be used then this field should be dropped from slide 5 and not added here.
- The `Pulse stretch (int)` field is needed here.  Default to 4 is fine.

# Slide 7

- Drop field `Number of samples` it is meaningless here
- Add field `Step Size (int)` with default of 1.

# Slide 8

- Increase default number of samples to 1000

# Slide 9

- In previous slides were calibration module is used you have a calphase input field but not here, if we should allow this to be a variable parameter it should be here as well.  Could you add this?
- How does `Points in dac range (odd)` transform to the actual trim points that are being used?  e.g. if I enter 5 what trim DAC values are used?  Why not just have a comma separated list of trim DAC values defaulting to `-63,0,63`?  This to me is easier to understand.
- Also it was unclear to me what the final goal of *this* was.  Did we decide to have this only taking the scurves followed by some "offline" analysis to derive a config that is tested in a later scurve run?  Maybe more a question for @mexanick.

# Slide 10

- Right now the routine that performs this does not take an input for the number of ADC reads (e.g. number of samples) to perform for each DAC value (it's presently hard coded to 100).  So perhaps this field should be dropped entirely.

# Slide 11

- Nice drop down here, I would set the defaults such that all boxes were checked *except* the following:
   - `CFG_CAL_DAC`, and
   - `CFG_VREF_ADC`

This would represent the most useful default behavior.  You mentioned it's possible to easily select all, is it also easy to deselect all? If not could you add a `deselect all` button?

# Slide 12

Could you add a field that specifies the ARM DAC points to use for the calibration which defaults to the following list: `17,20,23,25,30,40,50,60,70,80,100,125,150,175`?

bdorney reviewed (changes requested):
Changes on interface layout requested.

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,244 @@
+	  {CFG_THR_ZCC_DAC, "CFG_THR_ZCC_DAC"},
+	  {CFG_BIAS_PRE_VREF,"CFG_BIAS_PRE_VREF"},
+	  {CFG_VREF_ADC, "CFG_VREF_ADC"},
+	};
```
* Can all these various maps be `const`/`constexpr`?
  * They seem to simply be placeholders for templates for the UI display defaults.
* Some of them seem to be duplicated unnecessarily:
  * `m_dacScanTypeParams_label` and `m_dacScanTypeParams` could be merged to give a mapping:
    * `enum` -> `{label, min, max}`

bdorney commented:
Another question; this interface appears to let a user set these values and then execute a scan. Suppose I want to come back to a long scan some time later when it’s still running. How can I check the entered values at runtime? Through RCMS? And how can I check what the current position in the scan is?

cgalloni committed:
changing tab to 4 white spaces

mexanick commented:
> Another question; this interface appears to let a user set these values and then execute a scan. Suppose I want to come back to a long scan some time later when it’s still running. How can I check the entered values at runtime? Through RCMS? 

This is a good question. So far we don't have a well-established mechanism. The easiest thing to do is to disable form controls (read buttons, input fields etc.) but continue showing the screen. However it would be better to store the scan configuration in the DB. But this has to come later, after we manage to integrate the DB interactions for detector configurations 

> And how can I check what the current position in the scan is?

My idea was to implement a progress bar based on the range and its execution. Will be added when the routines itself will be attached.

bdorney commented:
> This is a good question. So far we don't have a well-established mechanism. The easiest thing to do is to disable form controls (read buttons, input fields etc.) but continue showing the screen. However it would be better to store the scan configuration in the DB. But this has to come later, after we manage to integrate the DB interactions for detector configurations

Disabling the form controls would be nice; I assume the running xdaq process would show this, i.e. if another user navigates to the hyperdaq page they can see this info as well.

I agree it would be good to have scan configuration in the DB as well.

> My idea was to implement a progress bar based on the range and its execution. Will be added when the routines itself will be attached.

Great!

cgalloni committed:
changed structure of parameters for DAC scans, added default button, other changes

cgalloni committed:
removed forgotten commented out lines

cgalloni commented:
Hi @bdorney,   please see the ansewrs in line.

> What is CalPulse phase here supposed to be? Is this the time between L1A and calpulse? Or is this the phase in the VFAT3 calibration unit? e.g. GEM_AMC.OH.OHX.GEB.VFATY.CFG_CAL_PHI?

This was intetended to be the pulse delay,  so it was changed to "Pulse delay" and set as default to 40

> Now slide specific comments.
> Slide 3
> 
>     A GBT Phase scan does not need a trigger source, drop this.
>     Add the following additional fields:
>         phase min (int) defaulting to 0,
>         phase max (int) defaulting to 14 (we've seen phase 15 is always bad, so no reason to include it)
>         step size (int) defaulting to 1
> Also the yellow bar says HELLOW WORLD. I guess this is not intended?

Changed

> Slide 4
> 
>     Change default ohMask to 0xfff since we should assume in P5 we will have 12 OH's per CTP7; or default it to 0xffc since on QC8 OH0 and OH1 are not used per CTP7. But the 0x3ff is a legacy from the slice test.

Changed to 0xffc for QC8.

> Slide 5
> 
>     Change default Throttle to 0; our use case in near term is entirely cosmic data taking or internal cal pulse and for cosmic I fear having a default of 100 will cause the scans to take 100 times as longer :D
>     Change default CFG_THR_ARM_DAC to at least 50 or 100. Having a default of 0 will probably risk to end up recording only noise.
> 
> If the calibration pulse radial button is selected then the Pulse Delay field (e.g. from slide 6) should be present here. It's meaningless for the particle option.
> 
> Here I assume Pulse Delay is the difference between the calpulse and the L1A and this should be configurable for a latency scan with the calibration pulse."
> 
> Slide 6
> 
>     What's the rational in including a field for CFG_THR_ARM_DAC on slide 5 for the latency scan but not here? Just wondering if we should be adding/dropping this; Is the use case the user configures with some known config in the DB? Or makes a config by hand and then makes a test? Or...? If a known config from the DB is to be used then this field should be dropped from slide 5 and not added here.
>     The Pulse stretch (int) field is needed here. Default to 4 is fine.

Changes implemented.  The CFG_THR_ARM_DAC  was masked form the interface as it should be loaded from DB. 


> Slide 9
> 
>     In previous slides were calibration module is used you have a calphase input field but not here, if we should allow this to be a variable parameter it should be here as well. Could you add this?

`Calphase` parameter was removed form the rest of the interface and here there is the Pulse Delay parameter.   


>    How does Points in dac range (odd) transform to the actual trim points that are being used? e.g. if I enter 5 what trim DAC values are used? Why not just have a comma separated list of trim DAC values defaulting to -63,0,63? This to me is easier to understand.
>     Also it was unclear to me what the final goal of this was. Did we decide to have this only taking the scurves followed by some "offline" analysis to derive a config that is tested in a later scurve run? Maybe more a question for @mexanick.

Implementd a string with {-63, 0, 63}.

> Slide 11
> 
>     Nice drop down here, I would set the defaults such that all boxes were checked except the following:
>         CFG_CAL_DAC, and
>         CFG_VREF_ADC
> 
> This would represent the most useful default behavior. You mentioned it's possible to easily select all, is it also easy to deselect all? If not could you add a deselect all button?

Yes, at the moment there are 3 buttons " DEFAULT" ,"SELECT  ALL" and "DESELECT ALL"

Slide 12

Could you add a field that specifies the ARM DAC points to use for the calibration which defaults to the following list: 17,20,23,25,30,40,50,60,70,80,100,125,150,175?

Done. I'll prepare a set of slides with the current layout.

cgalloni committed:
change of parameters type, adding jquery to gembase

cgalloni commented:
Hi @bdorney @jsturdy  @mexanick,
here is the new version of the calibration interface.  Please let me know if you have further comments. 
Thanks and cheers,

[CalibrationInterface3.pdf](https://github.com/cms-gem-daq-project/cmsgemos/files/3127515/CalibrationInterface3.pdf)

bdorney commented:
> CalibrationInterface3.pdf

I think slide 5 has the GBT scan interface instead of the latency scan interface by mistake?

bdorney commented:
> Hi @bdorney @jsturdy @mexanick,
> here is the new version of the calibration interface. Please let me know if you have further comments.
> Thanks and cheers,
> 
> [CalibrationInterface3.pdf](https://github.com/cms-gem-daq-project/cmsgemos/files/3127515/CalibrationInterface3.pdf)

So for every slide except slide 5 (latency scan) the interface looks great from my side.  No further comments there.  If you can re-upload a version that has slide 5 for latency scan I'll take a look at that later.

cgalloni commented:
Hi @bdorney , 
yes, you are right , I made a mistake with the cut and paste. Apologies.
Here the correct one:
[CalibrationInterface3.pdf](https://github.com/cms-gem-daq-project/cmsgemos/files/3128158/CalibrationInterface3.pdf)

bdorney commented:
> Hi @bdorney ,
> yes, you are right , I made a mistake with the cut and paste. Apologies.
> Here the correct one:
> [CalibrationInterface3.pdf](https://github.com/cms-gem-daq-project/cmsgemos/files/3128158/CalibrationInterface3.pdf)

Thanks looks good. No objections from my side.

mexanick reviewed (dismissed):
Looks good to me

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,224 @@
+            struct dacScanTypeFeature{
+                std::string label;
+                std::map<std::string, uint32_t> range;
+            };
```
OK, let me be more explicit.
For the usage as I've understood it here, the type that would be most appropriate would be a `union`
```c++
union dacFeature {
   std::string label;
   uint16_t min;
   uint16_t max;
};
```

jsturdy replied:
As @cgalloni found, in fact `union` was *not* the appropriate type for this usage, but a `struct` with two additional fields rather than a `map` was.

jsturdy commented on gemcalibration/include/gem/calib/Calibration.h:
```diff
@@ -0,0 +1,224 @@
+
+            xdata::Integer m_nShelves;
+
+            log4cplus::Logger m_logger;
```
In a previous iteration, this line had a comment to be removed, is that not the case?

jsturdy commented on gemcalibration/include/gem/calib/CalibrationWeb.h:
```diff
@@ -0,0 +1,135 @@
+
+#endif  // GEM_CALIB_CALIBRATIONWEB_H
+
+// Local Variables:
```
This type of stuff is generally frowned upon in this repository, please remove

jsturdy commented on gemcalibration/src/common/Calibration.cc:
```diff
@@ -0,0 +1,172 @@
+ 
+}
+
+bool gem::calib::Calibration::isGEMApplication(const std::string& classname) const
```
I don't think this function is relevant for this application

jsturdy commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,307 @@
+    std::map<std::string, std::string> t_parameters = dynamic_cast<gem::calib::Calibration*>(p_gemApp)->m_scanParams.find(m_calType)->second; 
+
+    for (auto parameter: t_parameters) {
+        if (dynamic_cast<gem::calib::Calibration*>(p_gemApp)->m_scanParamsNonForm.find(parameter.first) !=dynamic_cast<gem::calib::Calibration*>(p_gemApp)->m_scanParamsNonForm.end()) {
```
As you use this conversion multiple times in this function (and in other places where you are doing multiple casts), probably better to create a local variable, e.g.,
```c++
auto gemAppLocal = dynamic_cast<gem::calib::Calibration*>(p_gemApp);
```

cgalloni committed:
changing to struct for DAC scan parameters,removed commented out lines

changing to struct for DAC scan parameters

removed commented out lines

jsturdy reviewed (changes requested):
Can you squash your last two commits into one (so the removal of commented lines, and the structural change that introduced those commented lines) are a single commit?

mexanick commented:
@jsturdy I guess all your comments has been addressed... Please check this

jsturdy reviewed (approved):
> @jsturdy I guess all your comments has been addressed... Please check this

Yes, all other issues I have with the code can be addressed in future improvements
